axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/InvoicingProjectService.java

```diff
@@ -453,6 +469,54 @@ public void fillLines(InvoicingProject invoicingProject, Project project) {
                 .filter(taskQueryBuilder.toString())
                 .bind(taskQueryMap)
                 .fetch());
+
+    Set<StockMoveLine> deliverySet = new HashSet<>();
+    processDeliveredSaleOrderLines(deliverySet, saleOrderLineList);
+    invoicingProject.getDeliverySet().addAll(deliverySet);
+  }
+
+  public void processDeliveredSaleOrderLines(
+      Set<StockMoveLine> deliverySet, List<SaleOrderLine> saleOrderLineList) {
+    for (SaleOrderLine saleOrderLine : saleOrderLineList) {
+      if (shouldProcessSaleOrderLine(saleOrderLine)) {
+        processStockMoveLines(deliverySet, saleOrderLine);
+      }
+    }
+  }
+
+  private boolean shouldProcessSaleOrderLine(SaleOrderLine saleOrderLine) {
+    return saleOrderLine.getInvoicingModeSelect()
+            == SaleOrderLineSaleRepository.INVOICING_MODE_ON_DELIVERY
+        && saleOrderLine.getDeliveryState() == SaleOrderLineRepository.DELIVERY_STATE_DELIVERED;
+  }
+
+  private void processStockMoveLines(Set<StockMoveLine> deliverySet, SaleOrderLine saleOrderLine) {
+    List<StockMoveLine> stockMoveLineList = findRelevantStockMoveLines(saleOrderLine);
+
+    for (StockMoveLine stockMoveLine : stockMoveLineList) {
+      if (isStockMoveLineEligible(stockMoveLine)) {
+        deliverySet.add(stockMoveLine);
+      }
+    }
+  }
+
+  private List<StockMoveLine> findRelevantStockMoveLines(SaleOrderLine saleOrderLine) {
+    return Beans.get(StockMoveLineRepository.class)
+        .all()
+        .filter(
+            "self.saleOrderLine = :saleOrderLine AND self.stockMove IS NOT NULL "
+                + "AND self.stockMove.statusSelect = :statusSelect")
+        .bind("saleOrderLine", saleOrderLine)
+        .bind("statusSelect", StockMoveRepository.STATUS_REALIZED)
+        .fetch();
+  }
+
+  private boolean isStockMoveLineEligible(StockMoveLine stockMoveLine) {
```

Comment:
Use protected instead of private. Overring methods will be easier in the future.